### PR TITLE
guac integration

### PR DESCRIPTION
**Pull request by kingoflolz**

### CODE REVIEW

File: exit_db/migrations/2018-07-29-124240_add_eth_addr/down.sql [hunk 1]
```diff
@@ -0,0 +1,20 @@
+ALTER TABLE clients RENAME TO clients_old;
+
+CREATE TABLE clients
+(   mesh_ip VARCHAR NOT NULL PRIMARY KEY,
+    wg_pubkey VARCHAR NOT NULL,
+    wg_port VARCHAR NOT NULL,
+    luci_pass VARCHAR NOT NULL,
```
Comment (jkilpatr) [inline exit_db/migrations/2018-07-29-124240_add_eth_addr/down.sql]: go ahead and remove this.

File: rita/src/rita_common/dashboard/mod.rs [hunk 2]
```diff
@@ -42,16 +48,22 @@ impl Handler<GetOwnInfo> for Dashboard {
...
-                .send(GetOwnBalance {})
-                .from_err()
-                .and_then(|res| {
-                    Ok(OwnInfo {
-                        balance: res?,
-                        version: env!("CARGO_PKG_VERSION").to_string(),
-                    })
-                }),
-        )
+        #[cfg(feature = "guac")]
```
Comment (jkilpatr) [inline rita/src/rita_common/dashboard/mod.rs]: this is cleaner than that it replaces, we should try and use this pattern more often.

File: rita/src/rita_common/debt_keeper/mod.rs [hunk 3]
```diff
@@ -87,6 +89,45 @@ impl Handler<PaymentReceived> for DebtKeeper {
...
+                        "[{}]:{}",
+                        id.mesh_ip,
+                        SETTING.get_network().guac_contact_port
+                    ),
+                }));
+
+                ctx.spawn(
+                    PaymentController::from_registry()
+                        .send(Withdraw(id.eth_address))
+                        .into_actor(self)
```
Comment (jkilpatr) [inline rita/src/rita_common/debt_keeper/mod.rs]: why do we need to turn these futures into actors?

File: bounty_hunter/src/schema.rs [hunk 4]
```diff
@@ -1,7 +0,0 @@
-table! {
-    status (ip) {
-        ip -> Text,
-        mac -> Text,
-        balance -> Text,
-    }
-}
```
[the remark this answers is not in the document]
Comment (kingoflolz) [reply bounty_hunter/src/schema.rs]: Bounty hunter is going to be moved into guac eventually, and now the balances for the integration tests are done via requests to each individual node

File: integration-tests/rita.py [hunk 5]
```diff
@@ -946,6 +905,8 @@ def main():
         if choice != 'y':
             sys.exit(0)
 
+    time.sleep(10)
```
[the remark this answers is not in the document]
Comment (kingoflolz) [reply integration-tests/rita.py]: We don't actually check convergence within the exit tunnels, and I was debugging some exit billing issues so I commented out the other tests. I'll remove this for now but we should think about adding convergence tests for the exit tunnels

File: rita/src/rita_common/dashboard/mod.rs [hunk 6]
```diff
@@ -1,9 +1,15 @@
 use actix::prelude::*;
+use actix::registry::SystemService;
 
 use failure::Error;
+#[cfg(not(feature = "guac"))]
+use futures;
+
+#[cfg(feature = "guac")]
```
[the remark this answers is not in the document]
Comment (kingoflolz) [reply rita/src/rita_common/dashboard/mod.rs]: We only need this when we use futures combinators, and we don't use it when we don't have guac enabled

File: settings/src/lib.rs [hunk 7]
```diff
@@ -66,6 +66,10 @@ lazy_static! {
     static ref KI: Box<KernelInterface> = Box::new(LinuxCommandRunner {});
 }
 
+fn default_guac_contact_port() -> u16 {
+    4874
```
[the remark this answers is not in the document]
Comment (kingoflolz) [reply settings/src/lib.rs]: perhaps, we still need to work out how settings works across guac and rita, but right now I have rita configure guac (so guac doesn't have its own config file)

File: settings/src/lib.rs [hunk 8]
```diff
@@ -77,9 +81,10 @@ pub struct NetworkSettings {
     /// Port on which rita starts the per hop tunnel handshake on (needs to be constant across an
     /// entire althea deployment)
     pub rita_hello_port: u16,
-    /// Port on which rita contacts other althea nodes over the mesh (needs to be constant across an
+    /// Port on which guac contacts other althea nodes over the mesh (needs to be constant across an
     /// entire althea deployment)
-    pub rita_contact_port: u16,
+    #[serde(default = "default_guac_contact_port")]
```
[the remark this answers is not in the document]
Comment (kingoflolz) [reply settings/src/lib.rs]: Not sure what you mean here, but afaik the only way to create a default value is to add a function and set that as a serde default macro attr

File: bounty_hunter/src/schema.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (drozdziak1) [reply bounty_hunter/src/schema.rs]: Makes sense, thanks

File: integration-tests/rita.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (drozdziak1) [reply integration-tests/rita.py]: Yeah, I'll actually be overhauling a lot of what we know as `rita.py` in the forseeable future. Agreed about the exit tunnels :+1:

File: rita/src/rita_common/dashboard/mod.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (drozdziak1) [reply rita/src/rita_common/dashboard/mod.rs]: Makes sense

File: settings/src/lib.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (drozdziak1) [reply settings/src/lib.rs]: Okay, it's just other default stuff in `NetworkSettings` is handled in the `Default` impl

**Comment by mpapierski** [comment]
@kingoflolz can you rebase this change against master?

File: Cargo.toml [hunk 9]
```diff
@@ -11,12 +11,12 @@ members = ["althea_kernel_interface", "bounty_hunter", "settings", "clu", "exit_
 
 [profile.release]
 opt-level = "z"
-debug=true
```
[the remark this answers is not in the document]
Comment (mpapierski) [reply Cargo.toml]: In current master this is already gone from release profile.

File: Cargo.toml [hunk 10]
```diff
@@ -6,17 +6,25 @@ authors = ["Stan Drozd <drozdziak1@gmail.com>"]
...
 lto = true
 codegen-units = 1
 incremental = false
 
 [patch.crates-io]
+ethereum-types = {git="https://github.com/paritytech/primitives.git"}
```
Comment (mpapierski) [inline Cargo.toml]: What about pinning this to a specific commit?

File: rita/src/client.rs [hunk 11]
```diff
@@ -166,33 +166,33 @@ fn main() {
...
 
     assert!(rita_common::debt_keeper::DebtKeeper::from_registry().connected());
-    assert!(rita_common::payment_controller::PaymentController::from_registry().connected());
     assert!(rita_common::tunnel_manager::TunnelManager::from_registry().connected());
     assert!(rita_common::http_client::HTTPClient::from_registry().connected());
     assert!(rita_common::traffic_watcher::TrafficWatcher::from_registry().connected());
+
     assert!(rita_client::exit_manager::ExitManager::from_registry().connected());
 
+    #[cfg(feature = "guac")]
```
Comment (mpapierski) [inline rita/src/client.rs]: Maybe its a matter of taste but isn't this more readable for conditional compiling of code?

```rust
if cfg!(feature = "guac") {
  // foo
}
```
?

File: rita/src/rita_common/dashboard/mod.rs (hunk 2 above)
Comment (mpapierski) [reply rita/src/rita_common/dashboard/mod.rs]: BTW for blocks of feature dependent code wouldn't it be more readable to do `if cfg!(...)` instead?

File: rita/src/client.rs (hunk 11 above)
Comment (kingoflolz) [reply rita/src/client.rs]: The difference is that this way its cut out at compile time instead of at runtime

File: Cargo.toml (hunk 10 above)
Comment (kingoflolz) [reply Cargo.toml]: the punning is done in cargo.lock, and we check that into git, so it should be consistent

**Comment by mpapierski** [comment]
Hey @kingoflolz if you need any help with this just let me know I'll be happy to pick it up and merge in.

**Comment by mpapierski** [comment]
Closed because it became really out of date. I have a tracking issue for integrating this in #236.

**Pull request closed without merging**